src/riak_kv_qry_compiler.erl

```diff
@@ -1097,9 +1110,378 @@ modify_where_key(TupleList, Field, NewVal) ->
+            (_, or_, Acc) ->
+                {skip, Acc};
+            (_, Filter, Acc) ->
+                case lists:member(Filter, ElimLater) of
+                  true ->
+                      {eliminate, Acc};
+                  false ->
+                      {Filter, Acc}
+                end
+        end, [], Filter2),
+    lists:keystore(filter, 1, WhereProps, {filter,Filter3}).
+
+check_where_clause_is_possible_fold(DDL, _, {'=',{field,FieldName,_},{const,?SQL_NULL}} = F, Acc) ->
+    %% `IS NULL` filters
+    %% if a column is marked NOT NULL, but the WHERE clause has IS NULL for that
+    %% column then no results will ever be returned, so do not execute!
+    case is_field_nullable(FieldName, DDL) of
+        true ->
+            {F, Acc};
+        false ->
+            throw({error, {impossible_where_clause, << >>}})
+    end;
+check_where_clause_is_possible_fold(DDL, _, {'!=',{field,FieldName,_},{const,?SQL_NULL}} = F, Acc) ->
+    %% `NOT NULL` filters
+    %% if column is marked NOT NULL, and the WHERE clause has IS NOT NULL for
+    %% that column then we can eliminate it, and reduce the filter, maybe to
+    %% nothing which means leveldb would not have to decode the rows!
+    case is_field_nullable(FieldName, DDL) of
+        true ->
+            {F, Acc};
+        false ->
+            {eliminate, Acc}
+    end;
+check_where_clause_is_possible_fold(_, _, {'=',{field,FieldName,_},{const,EqVal}} = F, Acc) ->
+    case find_filter_check(FieldName, Acc) of
+        #filtercheck{'=' = F} ->
+            %% this filter has been specified twice so remove the second occurence
+            {eliminate, Acc};
+        #filtercheck{'>' = {_,_,{const,GtVal}} = GtF} = Check1 when GtVal < EqVal ->
+            %% query like `a = 10 AND a > 8` we can eliminate the greater than
+            %% clause because if a to be 10 it must also be greater than 8
+            Acc2 = append_to_eliminate_later(GtF, Acc),
+            Check2 = Check1#filtercheck{'>' = undefined, '=' = F},
+            {F, lists:keystore(FieldName, #filtercheck.name, Acc2, Check2)};
+        #filtercheck{'>=' = {_,_,{const,GteVal}} = GteF} = Check1 when GteVal < EqVal ->
+            %% query like `a = 10 AND a > 8` we can eliminate the greater than
+            %% clause because if a to be 10 it must also be greater than 8
+            Acc2 = append_to_eliminate_later(GteF, Acc),
+            Check2 = Check1#filtercheck{'>=' = undefined, '=' = F},
+            {F, lists:keystore(FieldName, #filtercheck.name, Acc2, Check2)};
+        #filtercheck{'<' = {_,_,{const,LtVal}}} when LtVal =< EqVal ->
+            %% query requires a column to be equal to a value AND less than that
+            %% value which is impossible
+            throw({error, {impossible_where_clause, << >>}});
+        #filtercheck{'<=' = {_,_,{const,LteVal}} = LteF} = Check1 when LteVal >= EqVal ->
+            %% query like `a = 10 AND a <= 10` the less than or equals can be
```

Comment:
There are more redundant cases, such as `a > 8 AND a > 10`. These redundancies are not truly part of the implied semantic of filtercheck. A different way of handling such redundancies is by having a phase following filtercheck that normalizing the filter ranges, eliminating fully contained filter ranges, i.e. {a, (10, infinity)} is fully contained by {a, (8, infinity)}.

While this change is not included in the scope of the work, despite the lack of total coverage of eliminating redundant filter cases, I do feel like it advances the product's UX and performance, reducing a good amount of redundant filter cases.